### PR TITLE
Add optimistic locking support to dynamo backend

### PR DESCRIPTION
**Pull request by rosstimothy**
The dynamo test was also updated to allow the table name to be overridden via an environment variable.

Contributes to #30416.

### CODE REVIEW

File: lib/backend/dynamo/dynamodbbk.go [hunk 1]
```diff
@@ -562,6 +572,47 @@ func (b *Backend) Delete(ctx context.Context, key []byte) error {
...
+	av, err := dynamodbattribute.MarshalMap(keyLookup{
+		HashKey:  hashKey,
+		FullPath: prependPrefix(key),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	input := dynamodb.DeleteItemInput{Key: av, TableName: aws.String(b.TableName)}
+	input.SetConditionExpression("#v = :rev")
+	input.SetExpressionAttributeNames(map[string]*string{"#v": aws.String("Revision")})
```
Comment (espadolini) [inline lib/backend/dynamo/dynamodbbk.go:603]: Doesn't this work? I'm not sure why we do this in `CompareAndSwap`.

```suggestion
	input.SetConditionExpression("Revision = :rev")
```

File: lib/backend/dynamo/dynamodbbk.go [hunk 2]
```diff
@@ -562,6 +572,47 @@ func (b *Backend) Delete(ctx context.Context, key []byte) error {
...
+	item.Revision = rev
+	return backend.NewLease(item), nil
+}
+
+// ConditionalDelete deletes item by key if the provided revision matches
+// the revision of the item in Dynamo.
+func (b *Backend) ConditionalDelete(ctx context.Context, key []byte, rev string) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
```
Comment (espadolini) [inline lib/backend/dynamo/dynamodbbk.go:592]: `create` (as called by `Create`, `Update` and `Put` and now `ConditionalUpdate`) doesn't check that the key is nonempty.

File: lib/backend/dynamo/dynamodbbk.go [hunk 3]
```diff
@@ -838,42 +881,55 @@ func trimPrefix(key string) []byte {
...
 	switch mode {
 	case modeCreate:
 		input.SetConditionExpression("attribute_not_exists(FullPath)")
 	case modeUpdate:
 		input.SetConditionExpression("attribute_exists(FullPath)")
 	case modePut:
+	case modeConditionalUpdate:
+		input.SetExpressionAttributeNames(map[string]*string{"#v": aws.String("Revision")})
+		input.SetExpressionAttributeValues(map[string]*dynamodb.AttributeValue{":rev": {S: aws.String(item.Revision)}})
+		input.SetConditionExpression("#v = :rev AND attribute_exists(FullPath)")
```
Comment (espadolini) [inline lib/backend/dynamo/dynamodbbk.go:918]: Same as in `ConditionalDelete`.

```suggestion
		input.SetExpressionAttributeValues(map[string]*dynamodb.AttributeValue{":rev": {S: aws.String(item.Revision)}})
		input.SetConditionExpression("Revision = :rev AND attribute_exists(FullPath)")
```

File: lib/backend/dynamo/dynamodbbk.go (hunk 2 above)
Comment (espadolini) [reply lib/backend/dynamo/dynamodbbk.go:592]: If anything we should allow deleting a key that's empty but not write it - right now we're in the weird position of allowing a Put with an empty key with no way to delete it 😬 

If we can distinguish between an empty but set FullPath and no FullPath (as we use that for existence), why not just allow the empty key?

File: lib/backend/dynamo/dynamodbbk.go [hunk 4]
```diff
@@ -632,22 +682,14 @@ const (
 	tableStatusMissing
 	tableStatusNeedsMigration
 	tableStatusOK
+	modeConditionalUpdate
```
Comment (espadolini) [inline lib/backend/dynamo/dynamodbbk.go:685]: Wrong `const` group

**Review by espadolini — APPROVED** [review]

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**